File: guests/swift/sliders.swift
// The sliders scene, Swift port — guests/rust/sliders.rs,
// tools/scenes/sliders.steps, docs/slider-plan.md.

import Foundation

struct Track: KayaGen {
    var name: String
    var level: Double
}

// The harness's own slider spelling (crates/kaya/src/harness.rs).
func spelled(_ v: Double) -> String {
    var s = String(format: "%.6f", v)
    while s.hasSuffix("0") { s.removeLast() }
    if s.hasSuffix(".") { s.removeLast() }
    return s
}

let app = KayaApp()
var commits = 0

app.build { tx in
    let levelText = tx.signal(.str("value: 50"))
    let commitText = tx.signal(.str("commits: 0"))
    let volumeText = tx.signal(.str("volume: 0.5"))
    let rowText = tx.signal(.str("row: none"))
    let pos = tx.signal(.f64(50.0))
    let tracks = trackCollection(tx)

    let root = tx.column {
        tx.label(bind: levelText)  // label#0
        tx.label(bind: commitText)  // label#1
        tx.label(bind: volumeText)  // label#2
        tx.label(bind: rowText)  // label#3
        let master = tx.slider(  // slider#0
            min: 0.0, max: 100.0, step: 5.0, tickSpacing: 25.0, bind: pos,
            onChange: { tx, v in tx.write(levelText, .str("value: \(spelled(v))")) },
            onCommit: { tx, _ in
                commits += 1
                tx.write(commitText, .str("commits: \(commits)"))
            })
        tx.setA11yId(master, "master")
        tx.setA11yLabel(master, "Level")
        let volume = tx.slider(  // slider#1
            min: 0.0, max: 1.0, value: 0.5, tickSpacing: 0.25,
            onChange: { tx, v in tx.write(volumeText, .str("volume: \(spelled(v))")) })
        tx.setA11yLabel(volume, "Volume")
        tx.button("reset") { tx in  // button#0
            // Must NOT come back as a value or a commit occurrence.
            tx.write(pos, .f64(25.0))
        }
        for row in tracks.rows {
            row.label(row.name)
            let level = row.slider(
                min: 0.0, max: 100.0, value: row.level, step: 10.0,
                onCommit: { tx, keys, v in
                    guard case .str(let key) = keys[0] else { return }
                    tx.write(rowText, .str("row \(key): \(spelled(v))"))
                })
            row.t.setA11yId(level, "level")
        }
    }
    tx.mount(root)

    tracks.insert(tx, .str("a"), Track(name: "a", level: 70.0))
    tracks.insert(tx, .str("b"), Track(name: "b", level: 20.0))
}

app.run()

File: guests/ocaml/sliders.ml
(* The sliders scene, OCaml port — guests/rust/sliders.rs,
   tools/scenes/sliders.steps, docs/slider-plan.md. *)

open Kaya_app

type track = { name : string; level : float } [@@deriving kaya_gen]

(* The harness's own slider spelling (crates/kaya/src/harness.rs). *)
let spelled v =
  let s = Printf.sprintf "%.6f" v in
  let last = ref (String.length s) in
  while !last > 0 && s.[!last - 1] = '0' do
    decr last
  done;
  if !last > 0 && s.[!last - 1] = '.' then decr last;
  String.sub s 0 !last

let () =
  let app = Kaya_app.create () in
  let commits = ref 0 in

  build app (fun () ->
      let level_text = signal_str ("value: 50") in
      let commit_text = signal_str ("commits: 0") in
      let volume_text = signal_str ("volume: 0.5") in
      let row_text = signal_str ("row: none") in
      let pos = signal_f64 (50.0) in
      let tracks = collection_of track_record in

      let root =
        column
          [
            label ~bind:level_text;                           (* label#0 *)
            label ~bind:commit_text;                          (* label#1 *)
            label ~bind:volume_text;                          (* label#2 *)
            label ~bind:row_text;                             (* label#3 *)
            slider ~a11y_id:"master" ~a11y_label:"Level"      (* slider#0 *)
              ~min:0.0 ~max:100.0 ~bind:pos ~step:5.0 ~tick_spacing:25.0
              ~on_change:(fun v ->
                write level_text (("value: " ^ spelled v)))
              ~on_commit:(fun _ ->
                incr commits;
                write commit_text
                  (Printf.sprintf "commits: %d" !commits));
            slider ~a11y_label:"Volume"                       (* slider#1 *)
              ~min:0.0 ~max:1.0 ~value:0.5 ~tick_spacing:0.25
              ~on_change:(fun v ->
                write volume_text (("volume: " ^ spelled v)));
            button ~text:"reset"                              (* button#0 *)
              ~on_click:(fun () ->
                (* Must NOT come back as a value or a commit occurrence. *)
                write pos (25.0));
            each (record_handle tracks) (fun () ->
                Tpl.(
                  row
                    [
                      label ~bind_field:track_name;
                      slider ~a11y_id:"level" ~bind_field:track_level
                        ~min:0.0 ~max:100.0 ~step:10.0
                        ~on_commit:(fun keys v ->
                          let key = key_text (List.hd keys) in
                          write row_text
                            (Printf.sprintf "row %s: %s" key (spelled v)));
                    ]
                    ()));
          ]
          ()
      in
      mount root;

      insert_record tracks (str_key "a") { name = "a"; level = 70.0 };
      insert_record tracks (str_key "b") { name = "b"; level = 20.0 });

  exit (run app)
